Improve the performance of super

eileencodes opened the pull request:
This PR improves the performance of `super` calls. While working on some
Rails optimizations jhawthorn discovered that `super` calls were slower
than expected.

The changes here do the following:

1) Adds a check for whether the call frame is not equal to the method
entry iseq. This avoids the `rb_obj_is_kind_of` check on the next line
which is quite slow. If the current call frame is equal to the method
entry we know we can't have an instance eval, etc.
2) Changes `FL_TEST` to `FL_TEST_RAW`. This is safe because we've
already done the check for `T_ICLASS` above.
3) Adds a benchmark for `T_ICLASS` super calls.
4) Note: makes a chage for `method_entry_cref` to use `const`.

On master the benchmarks showed that `super` is 1.76x slower. Our
changes improved the performance so that it is now only 1.36x slower.

Benchmark IPS:

```
ruby 3.0.0dev (2020-09-14T17:29:54Z master 28e60b0045) [x86_64-darwin19]
Warming up --------------------------------------
               super   248.792k i/100ms
         method call   434.392k i/100ms
Calculating -------------------------------------
               super      2.479M (± 1.3%) i/s -     12.440M in   5.018636s
         method call      4.354M (± 2.0%) i/s -     22.154M in   5.090333s

Comparison:
         method call:  4353993.7 i/s
               super:  2479084.2 i/s - 1.76x  (± 0.00) slower
```

With changes:

```
ruby 3.0.0dev (2020-09-14T19:18:30Z improve-performanc.. 1acc68eca0) [x86_64-darwin19]
Warming up --------------------------------------
               super   296.373k i/100ms
         method call   418.651k i/100ms
Calculating -------------------------------------
               super      3.090M (± 2.4%) i/s -     15.708M in   5.085938s
         method call      4.131M (± 1.5%) i/s -     20.933M in   5.068066s

Comparison:
         method call:  4131279.6 i/s
               super:  3090399.1 i/s - 1.34x  (± 0.00) slower
```

Ruby VM benchmarks also showed an improvement:

Existing `vm_super` benchmark`.

```
$ make benchmark ITEM=vm_super

|          |compare-ruby|built-ruby|
|:---------|-----------:|---------:|
|vm_super  |     32.045M|   34.361M|
|          |           -|     1.07x|
```

New `vm_iclass_super` benchmark:

```
$ make benchmark ITEM=vm_iclass_super

|          |compare-ruby|built-ruby|
|:---------|-----------:|---------:|
|vm_super  |      2.867M|    3.809M|
|          |           -|     1.33x|
```

This is the benchmark script used for the benchmark-ips benchmarks:

```ruby
require "benchmark/ips"

class Foo
  def zuper; end
  def top; end

  last_method = "top"

  ("A".."M").each do |module_name|
    eval <<-EOM
    module #{module_name}
      def zuper; super; end
      def #{module_name.downcase}
        #{last_method}
      end
    end
    prepend #{module_name}
    EOM
    last_method = module_name.downcase
  end
end

foo = Foo.new

Benchmark.ips do |x|
  x.report "super" do
    foo.zuper
  end

  x.report "method call" do
    foo.m
  end

  x.compare!
end
```

Co-authored-by: Aaron Patterson <tenderlove@ruby-lang.org>
Co-authored-by: John Hawthorn <john@hawthorn.email>

tenderlove commented on benchmark/vm_iclass_super.yml:
```diff
@@ -0,0 +1,20 @@
+
+  obj = C.new
+benchmark:
+  vm_super: obj.m
```
@eileencodes can you fix the benchmark name please?  I think we did locally, but it doesn't look pushed. 🙇🏻‍♂️

ko1 commented:
I'm not sure it can work on corner cases, for example using `define_method` to share the same iseq with different classes.

eileencodes commented:
@tenderlove I fixed the benchmark. I also updated the benchmarks in my commit message because it seems like change was introduced to Ruby recently that made it even faster. I'm going to search for which commit after I finish writing this comment.

@ko1 - Aaron and I looked at your concerns today and determined this change is safe.

The call cache is not a problem here because:

1) we have a type check so there's no issues with lambda's
2) `define_method` can only be called on classes in the same inheritance chain
3) we're not skipping invalidating the call cache, just skipping the is_a? check

```ruby
class A
  def bar; :bar; end 
end

class B < A 
  def bar; super; end 
end

class C
  define_method(:baz, A.instance_method(:bar))
end

p C.new.baz
=> ruby_scripts/ruby_file.rb:9:in `define_method': bind argument must be a subclass of A (TypeError)
	from ruby_scripts/ruby_file.rb:9:in `<class:C>'
	from ruby_scripts/ruby_file.rb:8:in `<main>'

```

tenderlove commented:
The iseq pointer comparison is [restricted to only methods iseq type methods](https://github.com/ruby/ruby/pull/3545/files#diff-347392b62c9edab3c8ab5fb0d71a012fR680) (*not* block methods). `define_method` only allows iseq methods to be defined on objects with the the same ancestors.  I think this means that if the current frame's iseq is the same as the defined method iseq, we can skip the `is_a?` check because the `is_a?` check was made at _method definition time_.

ko1 commented:
How about that?

```
module M
  def foo
    super
  end
end

class C0
  def foo() = :C0
end
class D0
  def foo() = :D0
end

class C1 < C0
  define_method(:foo, M.instance_method(:foo))
end
class D1 < D0
  define_method(:foo, M.instance_method(:foo))
end

p C1.new.foo
p D1.new.foo
```

eileencodes commented:
On master and on this branch I get:

```
:C0
:D0
```

On Ruby 2.7 I get:

```
ruby_scripts/ko1_example.rb:8: syntax error, unexpected '='
  def foo() = :C0
ruby_scripts/ko1_example.rb:11: syntax error, unexpected '='
  def foo() = :D0
ruby_scripts/ko1_example.rb:22: syntax error, unexpected end-of-input, expecting `end'
```

tenderlove commented:
@ko1 on master:

```
[aaron@tc-lan-adapter ~/g/ruby (master)]$ make runruby
./miniruby -I./lib -I. -I.ext/common  ./tool/runruby.rb --extout=.ext  -- --disable-gems ./test.rb
:C0
:D0
```

On this branch:

```
[aaron@tc-lan-adapter ~/g/ruby (improve-performance-of-super)]$ make runruby
./miniruby -I./lib -I. -I.ext/common  ./tool/runruby.rb --extout=.ext  -- --disable-gems ./test.rb
:C0
:D0
```

This change has no impact to the inline cache.  We found that [this line](https://github.com/ruby/ruby/pull/3545/files#diff-347392b62c9edab3c8ab5fb0d71a012fL3302) is a bottleneck.  We can skip that check if we compare iseq pointers.  That conditional is to check [instance_eval inside a method](https://github.com/ruby/ruby/blob/b717f73402dc6f2d6ba6a1d72d9c8543a4f2cd27/test/ruby/test_super.rb#L276-L283), and we can skip the check by comparing the control frame to the method definition.

@eileencodes this syntax is new in master, you need to compare master to this branch 😁

ko1 commented:
ok thank you for confirming! I don't have any problem.
